<a href="https://colab.research.google.com/github/brendanpshea/logic-prolog/blob/main/StatisticsWithWerewolves.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Statistics With Werewolves
### A Little More Logical | Brendan Shea, PhD
**Statistics** is the science of data. It involves the processes of collecting, analyzing, interpreting, and presenting numerical information. This field is essential for making informed  decisions based on data, allowing us to extract meaningful patterns and conclusions from a sea of numbers. By using statistics, we can transform raw data into useful insights, providing the ability to forecast trends, test hypotheses, and understand the world in a more data-driven way. In this chapter, statistics will be our tool to unravel the mysteries hidden in the High School Werewolf Dataset.

Our journey through statistics is centered around the High School Werewolf Dataset. In this (fictional) 2,000-student high school students, a number are secretly werewolves. This dataset includes a variety of information, from physical characteristics like height and eye color to academic and behavioral data like GPA and detentions. It's an ideal setting to explore statistical concepts because it combines everyday school life with the extraordinary element of werewolves. The dataset provides a rich, multifaceted context for learning how to analyze and interpret various types of data.

In statistical terms, the entire student body of 2,000 represents our **population**. The population is the complete set of data that we are interested in studying. However, studying a whole population is often impractical, so statisticians use a **sample** - a smaller, manageable part of the population that is (or is thought to be) representative of the whole. In our case, thesample is a 20-student 11th-grade class.

Throughout the chapter, we'll explore various questions that blend statistical analysis with the intriguing theme of werewolves. For example, we might ask if werewolf students have a different average height than their human counterparts, or if there are significant differences in full moon absences between the two groups. We'll also investigate whether eye color can predict werewolf status and explore if detention patterns suggest nocturnal activities typical of werewolves. These questions, while set in a fictional context, are designed to provide real-world insights into how statistics can be used to uncover hidden patterns and tell compelling stories with data.

### Getting to Know Our Data
In this section, we'll introduce you to two powerful tools that will help us explore the High School Werewolf Dataset: Google Colab and Pandas. Don't worry if you're new to programming or feel a bit apprehensive about it; these tools are user-friendly and designed to make your journey into data analysis both enjoyable and insightful.

**Google Colab** is an online platform that allows you to write and execute Python code through your browser. Think of it as a digital notebook that not only lets you write notes but also run code, all in one place. The best part? You don't need to install any software on your computer. All you need is an internet connection and a Google account.

Colab notebooks are interactive, meaning you can write a piece of code, run it, and see the results immediately. This makes it an excellent tool for learning, experimentation, and working with data, especially for beginners. You'll be using Google Colab to access, analyze, and visualize the werewolf dataset, learning hands-on how to manipulate and understand data.

**Pandas** is a library in Python, a programming language, that's specifically designed for data manipulation and analysis. Think of it as a powerful calculator that can not only perform basic arithmetic but also organize and analyze large amounts of data with ease.

With Pandas, we can read our High School Werewolf Dataset, which is stored in a CSV (Comma-Separated Values) file, a common format for storing tabular data. The dataset is available at this link: High School Werewolf Data. Pandas allows us to load this dataset into a DataFrame – a table that's very similar to a spreadsheet you might see in Excel. Once the data is in a DataFrame, we can perform various operations like calculating averages, filtering data based on conditions (like finding all werewolf students), and creating visualizations.

When we combine Google Colab and Pandas, we have a powerful setup for working with our dataset. You'll be able to write and run Python code in Colab to manipulate and analyze the dataset using Pandas, even if you've never written a line of code before.

So, let's start this adventure! We'll begin by opening our dataset in Google Colab and taking our first look at the data using Pandas. As we go through different statistical concepts, you'll see how these tools make data analysis accessible and engaging.


### Step 1: Loading the Dataset into a DataFrame

First, we need to load our dataset into a structure called a DataFrame, which Pandas uses to store and manipulate data in a tabular form. Here's how you can do it:

In [2]:
import pandas as pd

url = 'https://github.com/brendanpshea/logic-prolog/raw/main/high_school_werewolf_data.csv'
school_df = pd.read_csv(url)

You can run this cell by pressing the 'Play' button or use the shortcut Shift + Enter to run the cell. This will execute the code, import Pandas, and load the dataset into a DataFrame named school_df.

##Step 2: Viewing the First Few Rows of the Dataset

Once the dataset is loaded, it's a good practice to view the first few rows. This helps us get an initial feel for the data – the columns, the type of values, and so on. You can do this by using the `head()` function in Pandas. In a new cell, we can type this:

In [4]:
school_df.head()

Sex  Height EyeColor  FullMoonAbsence   GPA  WerewolfParents  \
0    Male   71.79     Blue                5  2.82                0   
1  Female   62.05     Blue                1  3.67                0   
2    Male   70.14    Brown                1  3.79                0   
3    Male   70.83    Green                0  2.62                0   
4    Male   70.68    Brown                2  2.54                0   

   Detentions  IsWerewolf Homeroom  
0           1        True     10-A  
1           1       False     10-A  
2           0       False     10-A  
3           0       False     10-A  
4           6       False     10-A

Running this command will display the first five rows of our dataset. Later, we'll be creating a subset of this to represent a particular "class." But for now, we'll be focusing on the school as a whole.

## Data Dictionary for High School Werewolf Dataset
A **data dictionary** is a document that explains the variables in a dataet.  The dataset simulates a 2,000 student high school with a twist: some students are werewolves. The dataset is designed for educational purposes, allowing students and teachers to explore statistical concepts in a fun, engaging manner.

-   **Sex:** A categorical variable indicating the gender of the student. Possible values are 'Male' and 'Female'.

-   **Height:** A continuous variable representing the student's height in inches. Heights follow a normal distribution. On average, male students are taller than female students, and werewolf students tend to be taller than their non-werewolf peers.

-   **EyeColor:** A categorical variable indicating the eye color of the student. Possible values are 'Brown', 'Blue', 'Green', 'Grey', and 'Yellow'. Yellow eyes are a unique trait found only among werewolves.

-   **FullMoonAbsence:** A discrete variable representing the number of days the student was absent after a full moon. This variable is normally distributed, with werewolves tending to be absent more on such days.

-   **GPA:** A continuous variable representing the student's Grade Point Average. This variable is normally distributed and is the same for werewolves and non-werewolves.

-   **WerewolfParents:** A discrete variable indicating the number of the student's parents who are werewolves. Possible values are 0, 1, or 2, with different base probabilities for werewolf and non-werewolf students.

-   **Detentions:** A discrete variable indicating the number of times the student has been in detention. This follows a Pareto distribution, implying that most students have few detentions, but a few have many. There is no difference in this variable between werewolves and non-werewolves.

-   **IsWerewolf:** A binary variable indicating whether the student is a werewolf or not. Possible values are True (werewolf) or False (non-werewolf).

## Measures of Central Tendency

In statistics, measures of central tendency are used to identify the center of a data set, giving us a representative value that defines the middle of the data distribution. These measures are crucial in summarizing a large set of data with a single value that represents the entire group. In this section, we'll explore three primary measures of central tendency: mean, median, and mode.

The **mean** is the most commonly known measure of central tendency. It is calculated by adding all the values in a data set and then dividing by the number of values. The mean provides a useful overall measure when the data is uniformly distributed without extreme values (outliers).

*Example:* To calculate the mean height of students in our dataset, add all the students' heights together and then divide by the total number of students. If five students have heights in inches of 60, 62, 65, 68, and 70, the mean height is (60 + 62 + 65 + 68 + 70) / 5 = 65 inches.

The **median** is the middle value in a data set when it's arranged in ascending or descending order. If there is an even number of observations, the median is the average of the two middle values. The median is particularly useful when dealing with data that have outliers, as it is not as affected by them as the mean.

*Example:* To find the median height, sort the heights and pick the middle one. If our heights are 60, 62, 65, 68, and 70 inches, the median is 65 inches (the third value). If there's an additional height of 66 inches, the median is the average of the two middle values: (65 + 66) / 2 = 65.5 inches.

The **mode** is the most frequently occurring value in a data set. A data set may have one mode, more than one mode, or no mode at all. The mode is especially useful for categorical data where we want to know which is the most common category.

*Example:* In determining the mode for eye color in our dataset, if 'Brown' occurs most frequently among the students, then 'Brown' is the mode. If 'Brown' and 'Blue' are equally common, the data set is bimodal, and both colors are modes.

### Why and How to Use Each Measure

Each measure of central tendency gives a different perspective on the data:

-   Use the mean for a quick, general understanding of the dataset, especially when the data distribution is symmetrical without outliers.
-   Use the median to find the middle of the dataset, especially when the data has outliers or is not symmetrically distributed.
-   Use the mode to understand the most common category or value in your dataset, particularly with categorical data.

Understanding these measures helps you analyze datasets like our High School Werewolf Dataset effectively. They provide a simple yet powerful way to summarize and interpret large amounts of data, offering insights that might not be immediately apparent.

## How Do Werewolves Differ From Non-Werewolves?

This section'll delve into the differences between werewolves and non-werewolves in our High School Werewolf Dataset. Our goal is to use Pandas, a powerful data analysis tool in Python, to explore measures of central tendency - specifically mean, median, and mode. These measures will help us understand the typical characteristics within each group, shedding light on how werewolves stand apart from their non-werewolf peers.

Pandas makes it easy to calculate statistical measures across different groups in a dataset. We'll be using its functionalities to group our data by the 'IsWerewolf' variable and then calculate the mean, median, and mode for various other variables. This approach will allow us to compare these central tendencies between werewolves and non-werewolves, providing insights into their unique traits.

### Calculating Means

The mean offers a general understanding of average values in each group. We can code it as follows:

In [10]:
mean_values = school_df.groupby('IsWerewolf').mean(numeric_only=True)
mean_values

Height  FullMoonAbsence      GPA  WerewolfParents  Detentions
IsWerewolf                                                                  
False       66.686913            1.136  3.10658         0.069333    7.747333
True        68.297480            2.382  3.11778         0.274000    6.784000

The mean values we've obtained from the High School Werewolf Dataset provide some interesting insights into the differences and similarities between werewolf and non-werewolf students. Let's break down what each of these means tells us:

1.  Height

    -   Non-Werewolves: The average height is approximately 66.69 inches.
    -   Werewolves: The average height is higher, around 68.30 inches.
    -   This suggests that werewolf students are, on average, taller than their non-werewolf peers. This finding aligns with the typical lore of werewolves being larger or more imposing figures.
2.  FullMoonAbsence

    -   Non-Werewolves: On average, non-werewolf students are absent about 1.14 days after a full moon.
    -   Werewolves: Werewolf students have a higher average absence, around 2.38 days.
    -   Werewolf students tend to be absent more often after a full moon, which could be humorously attributed to their nocturnal activities during this time. This is a fun way to demonstrate how data can sometimes align with fictional narratives.
3.  GPA

    -   Non-Werewolves: The average GPA is approximately 3.11.
    -   Werewolves: Similarly, the average GPA for werewolves is about 3.12.
    -   There is virtually no difference in the average academic performance between werewolf and non-werewolf students. This suggests that werewolf status does not impact academic abilities or achievements.
4.  WerewolfParents

    -   Non-Werewolves: On average, the number of werewolf parents for non-werewolf students is close to 0 (0.069).
    -   Werewolves: Werewolf students have a higher average, about 0.274.
    -   Werewolf students are more likely to have werewolf parents, which makes sense within the context of werewolf lore. This variable offers a playful nod to the idea of werewolf traits being hereditary.
5.  Detentions

    -   Non-Werewolves: The average number of detentions is approximately 7.75.
    -   Werewolves: The average is slightly lower for werewolves, around 6.78.
    -   Contrary to what one might expect from mythical portrayals of werewolves as unruly, werewolf students do not have more detentions than their non-werewolf counterparts. In fact, they have slightly fewer on average, which challenges the stereotype of werewolves as more prone to mischief.

In summary, these mean values offer a blend of expected and unexpected insights. While some findings, like the taller average height of werewolves, align with popular werewolf myths, others, like the similar academic performance and lower detention rates, challenge these stereotypes. This analysis exemplifies how statistical measures can be used to explore and interpret data, whether in real-world scenarios or imaginative contexts like our High School Werewolf Dataset.

### Medians
Now, let's do the same thing for medians:

In [12]:
median_values = school_df.groupby('IsWerewolf').median(numeric_only=True)
median_values

Height  FullMoonAbsence   GPA  WerewolfParents  Detentions
IsWerewolf                                                            
False       66.850              1.0  3.13              0.0         2.0
True        68.345              2.0  3.17              0.0         2.0

The median values from the High School Werewolf Dataset offer another perspective on the data, particularly highlighting where and why they differ from the mean values we previously examined. Let's analyze these medians:

1.  Height

    -   Non-Werewolves: The median height is 66.85 inches.
    -   Werewolves: The median height is 68.345 inches.
    -   Similar to the mean, the median height for werewolves is greater than for non-werewolves, reaffirming that werewolves are generally taller. The closeness of the median to the mean suggests a fairly symmetric distribution of height within both groups.
2.  FullMoonAbsence

    -   Non-Werewolves: The median absence after a full moon is 1.0 day.
    -   Werewolves: For werewolves, it's 2.0 days.
    -   The median values are lower than the mean, especially for werewolves. This indicates that while the average (mean) absence for werewolves is pulled up by a few students with very high absences, the majority of werewolf students (as reflected in the median) miss fewer days.
3.  GPA

    -   Non-Werewolves: The median GPA is 3.13.
    -   Werewolves: The median GPA is 3.17.
    -   The median GPAs are very close to the means, suggesting a symmetric distribution of GPA scores among both werewolves and non-werewolves, with no significant outliers affecting the average.
4.  WerewolfParents

    -   Non-Werewolves and Werewolves: The median in both cases is 0.0.
    -   This median aligns with the mean, indicating that the majority of students, whether werewolves or not, do not have werewolf parents.
5.  Detentions

    -   Non-Werewolves and Werewolves: The median number of detentions is 2.0 for both groups.
    -   The median detentions are much lower than the mean, especially for non-werewolves. This suggests that while the average number of detentions is influenced by a small number of students with very high detention counts (as seen in the mean), most students have a lower number of detentions. This disparity between the mean and median indicates a right-skewed distribution: most students have fewer detentions, but a few outliers with high numbers of detentions increase the mean.

**Why Medians Differ From Means.** Medians are less affected by outliers and skewed distributions than means. In cases where the data is not symmetrically distributed, or where there are extreme values (outliers), the median can provide a more accurate representation of the 'typical' data point than the mean. This is evident in our analysis of FullMoonAbsence and Detentions, where the medians offer a different view of the data, suggesting that most students' experiences differ from what the mean might imply due to the influence of outliers.

### Modes
The **mode** is generally used for **categorical** data (such as eye color) as opposed to **numerical** data (such as GPA, height, etc). Let's take a look at how eye color is distributed.

In [17]:
mode_values = school_df.groupby('IsWerewolf')['EyeColor'].value_counts()
mode_values

IsWerewolf  EyeColor
False       Brown       577
            Blue        452
            Green       318
            Grey        153
True        Brown       133
            Blue        131
            Yellow      103
            Green        83
            Grey         50
Name: EyeColor, dtype: int64

The mode, as the most frequently occurring value in a dataset, is particularly useful for analyzing categorical data. In this case, we're looking at the distribution of eye color among werewolf and non-werewolf students. This analysis not only tells us which eye color is most common in each group but also highlights the importance of considering proportions when interpreting categorical data.

Analysis of Eye Color Modes

From the `value_counts()` output, we can see the following:

-   Non-Werewolves:
    -   Most common eye color is Brown (577 students).
    -   Followed by Blue (452 students), Green (318 students), and Grey (153 students).
-   Werewolves:
    -   Most common eye color is still Brown (133 students), closely followed by Blue (131 students).
    -   Notably, Yellow eyes, a unique trait among werewolves, occur significantly (103 students).
    -   Other colors like Green (83 students) and Grey (50 students) are also present but less frequent.

Importance of Proportions in Categorical Data

When analyzing categorical data like eye color, it's important to consider proportions in addition to the raw counts. This helps us understand the distribution of categories within the context of the entire group. For instance, while Brown is the most common eye color among both werewolves and non-werewolves, the proportion of Yellow eyes is notably high among werewolves, a unique feature not seen in non-werewolves.

To gain a clearer understanding of these proportions, you can produce a table that calculates the percentage of each eye color within werewolf and non-werewolf groups. Here's how you can do it in Pandas:

In [18]:
eye_color_proportions = school_df.groupby('IsWerewolf')['EyeColor'].value_counts(normalize=True) * 100
eye_color_proportions

IsWerewolf  EyeColor
False       Brown       38.466667
            Blue        30.133333
            Green       21.200000
            Grey        10.200000
True        Brown       26.600000
            Blue        26.200000
            Yellow      20.600000
            Green       16.600000
            Grey        10.000000
Name: EyeColor, dtype: float64